server/src/internalClusterTest/java/org/elasticsearch/indices/recovery/SnapshotBasedIndexRecoveryIT.java

```diff
@@ -0,0 +1,752 @@
+
+        assertThat(repository.totalBytesRead.get(), is(equalTo(0L)));
+
+        assertDocumentsAreEqual(indexName, numDocs + docsIndexedAfterSnapshot);
+    }
+
+    public void testRecoveryIsCancelledAfterDeletingTheIndex() throws Exception {
+        updateSetting(INDICES_RECOVERY_MAX_CONCURRENT_SNAPSHOT_FILE_DOWNLOADS.getKey(), "1");
+
+        try {
+            String sourceNode = internalCluster().startDataOnlyNode();
+            String indexName = randomAlphaOfLength(10).toLowerCase(Locale.ROOT);
+            createIndex(indexName,
+                Settings.builder()
+                    .put(IndexMetadata.SETTING_NUMBER_OF_SHARDS, 1)
+                    .put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, 0)
+                    .put(MergePolicyConfig.INDEX_MERGE_ENABLED, false)
+                    .put(IndexService.GLOBAL_CHECKPOINT_SYNC_INTERVAL_SETTING.getKey(), "1s")
+                    .put("index.routing.allocation.require._name", sourceNode)
+                    .build()
+            );
+
+            int numDocs = randomIntBetween(3000, 5000);
+            indexDocs(indexName, numDocs, numDocs);
+
+            String repoName = "repo";
+            createRepo(repoName, "fs");
+            createSnapshot(repoName, "snap", Collections.singletonList(indexName));
+
+            String targetNode = internalCluster().startDataOnlyNode();
+            assertAcked(
+                client().admin().indices().prepareUpdateSettings(indexName)
+                    .setSettings(Settings.builder()
+                        .put("index.routing.allocation.require._name", targetNode)).get()
+            );
+
+            MockTransportService targetMockTransportService =
+                (MockTransportService) internalCluster().getInstance(TransportService.class, targetNode);
+
+            CountDownLatch recoverSnapshotFileRequestReceived = new CountDownLatch(1);
+            CountDownLatch respondToRecoverSnapshotFile = new CountDownLatch(1);
+            AtomicInteger numberOfRecoverSnapshotFileRequestsReceived = new AtomicInteger();
+            targetMockTransportService.addRequestHandlingBehavior(PeerRecoveryTargetService.Actions.RESTORE_FILE_FROM_SNAPSHOT,
+                (handler, request, channel, task) -> {
+                    numberOfRecoverSnapshotFileRequestsReceived.incrementAndGet();
+                    recoverSnapshotFileRequestReceived.countDown();
+                    respondToRecoverSnapshotFile.await();
+                    handler.messageReceived(request, channel, task);
+                }
+            );
+
+            recoverSnapshotFileRequestReceived.await();
+
+            assertAcked(client().admin().indices().prepareDelete(indexName).get());
+
+            respondToRecoverSnapshotFile.countDown();
+
+            assertThat(indexExists(indexName), is(equalTo(false)));
+            assertThat(numberOfRecoverSnapshotFileRequestsReceived.get(), is(equalTo(1)));
```

Comment:
Can we put this validation into the request handling behavior? Since there is no waiting after the latch count down, we could risk not seeing that it actually continues to send download file requests?